src/lib/foreign-field.ts

```diff
@@ -0,0 +1,496 @@
+ *
+ * Internally, a foreign field element is represented as three native field elements, each of which
+ * represents a limb of 88 bits. Therefore, being a valid foreign field element means that all 3 limbs
+ * fit in 88 bits, and the foreign field element altogether is smaller than the modulus p.
+ * With default parameters, new `ForeignField` elements introduced in provable code are automatically
+ * constrained to be valid on creation.
+ *
+ * However, optimized code may want to forgo these automatic checks because in some
+ * situations they are redundant. Skipping automatic validity checks can be done
+ * by passing the `unsafe: true` flag:
+ *
+ * ```ts
+ * class UnsafeField extends createForeignField(17n, { unsafe: true }) {}
+ * ```
+ *
+ * You then often need to manually add validity checks:
+ * ```ts
+ * let x: UnsafeField;
+ * x.assertValidElement(); // prove that x is a valid foreign field element
+ * ```
+ *
+ * @param modulus the modulus of the finite field you are instantiating
+ * @param options
+ * - `unsafe: boolean` determines whether `ForeignField` elements are constrained to be valid on creation.
+ */
+function createForeignField(modulus: bigint, { unsafe = false } = {}) {
+  const p = modulus;
+  const pMl = ForeignFieldConst.fromBigint(p);
+
+  if (p <= 0) {
+    throw Error(`ForeignField: expected modulus to be positive, got ${p}`);
+  }
+  if (p > foreignFieldMax) {
+    throw Error(
+      `ForeignField: modulus exceeds the max supported size of 2^${foreignFieldMaxBits}`
+    );
+  }
+
+  let sizeInBits = p.toString(2).length;
+
+  class ForeignField {
+    static modulus = p;
+    value: ForeignFieldVar;
+
+    /**
+     * Create a new {@link ForeignField} from a bigint, number, string or another ForeignField.
+     *
+     * ```ts
+     * let x = new ForeignField(5);
+     * ```
+     */
+    constructor(x: ForeignField | ForeignFieldVar | bigint | number | string) {
+      if (x instanceof ForeignField) {
+        this.value = x.value;
+        return;
+      }
+      // ForeignFieldVar
+      if (Array.isArray(x)) {
+        this.value = x;
+        return;
+      }
+      // constant
+      this.value = ForeignFieldVar.fromBigint(mod(BigInt(x), p));
+    }
+
+    /**
+     * Coerce the input to a {@link ForeignField}.
+     */
+    static from(x: ForeignField | ForeignFieldVar | bigint | number | string) {
+      if (x instanceof ForeignField) return x;
+      return new ForeignField(x);
+    }
+
+    /**
+     * Checks whether this field element is a constant.
+     *
+     * See {@link FieldVar} tp understand constants vs variables.
+     */
+    isConstant() {
+      let [, ...limbs] = this.value;
+      return limbs.every(FieldVar.isConstant);
+    }
+
+    /**
+     * Convert this field element to a constant.
+     *
+     * See {@link FieldVar} to understand constants vs variables.
+     *
+     * **Warning**: This function is only useful in `Provable.witness()` or `Provable.asProver()` blocks,
```

Comment:
Nit: Is it possible to add a link to `Provable.witness` or `Provable.asProver` in the doc comments here?